### DISCUSSION TITLE
Where does GitHub Copilot processing happen?

### DISCUSSION THREAD

**Discussion by gansaikhanshur** [community/community#52280] [Copilot Conversations]
### Select Topic Area

Question

### Body

Does it upload my code to the cloud and then makes suggestions or does it somehow happen on a local device? Where can I find documentation on how it makes the processing?

**Comment by SMelidoni** [comment]
Hi @gansaikhanshur 👋🏽

GitHub Copilot processes and generates suggestions in the cloud. When you're using Copilot, your code is sent to GitHub's servers, where the AI model generates suggestions based on the context of your code. This means that your code, or parts of it, are transmitted over the internet to GitHub's infrastructure.

You can find more information about GitHub Copilot, including how it processes data, in the [GitHub Copilot Documentation](https://docs.github.com/en/copilot).

I hope this answers your question. If you need any further information or have any concerns, please let me know.

**Reply by gansaikhanshur** [reply]
Hi @SMelidoni! 

Thank you for your reply! How about for enterprise/business uses? I'm sure some companies wouldn't like the idea of their code being uploaded on GitHub servers for processing, how does GitHub copilot ensures privacy and security? 

Is it possible to host GitHub copilot locally?

**Reply by SMelidoni** [reply]
Hi!

GitHub Copilot currently only runs on GitHub servers and does not offer a local hosting option. However, in terms of security and privacy, GitHub Copilot analyses the code entered by developers to generate suggestions, and this information is collected for improvement purposes. GitHub manages the collected data strictly. They stated that:

_'Code snippets data is only transmitted in real-time to return Suggestions, and is discarded once a Suggestion is returned. Copilot for Business does not retain any Code Snippets Data.'_ (I have put the link to this at the bottom of this response)

For enterprise or business users, you should review the information protection policy and terms of service for GitHub Copilot before starting to use it. This will allow you to ensure that your company's security policies align with GitHub Copilot's security and privacy guidelines.

Additionally, companies can limit the scope of usage by using GitHub Copilot excluding confidential code or only using it for non-sensitive projects. It's important to note that the code you're working on will be sent to GitHub's servers for processing, so you need to weigh the risks and benefits according to your organisation's policies.

You can also find more information within this GitHub page as mentioned earlier for [GitHub Copilot For Business Privacy Statement](https://docs.github.com/en/site-policy/privacy-policies/github-copilot-for-business-privacy-statement)

**Reply by yousafe007** [reply]
Could you please update the link provided above?

**Reply by gitstua** [reply]
Another great resource to answer your questions are more is the [Copilot Trust Center](https://resources.github.com/copilot-trust-center/)

**Comment by bammermann** [comment]
Hey,

Updated Link would be: https://resources.github.com/copilot-trust-center/#privacy

> Prompts only persist long enough to return a response from GitHub’s Azure tenant and are then immediately discarded. Prompts are never stored; all processing is ephemeral and takes place in memory.

**Comment by mohitjakhotra** [comment]
If the processing is happening on the cloud then why does Macbook air m1 heats so much when using copilot in vs code as if it is running locally ?
